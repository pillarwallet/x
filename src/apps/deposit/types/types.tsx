export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
