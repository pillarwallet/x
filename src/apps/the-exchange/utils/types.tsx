import { BridgingProvider } from '@etherspot/data-utils/dist/cjs/sdk/data/constants';
import { Route } from '@lifi/sdk';
import { Hex } from 'viem';

export enum CardPosition {
  SWAP = 'SWAP',
  RECEIVE = 'RECEIVE',
}

export type SwapType = {
  fromAmount: number;
  fromTokenAddress: string;
  fromChainId: number;
  fromTokenDecimals: number;
  toTokenAddress: string;
  toChainId: number;
  toTokenDecimals: number;
  slippage: number;
  fromAccountAddress?: string;
  provider?: BridgingProvider;
};

export type SwapOffer = {
  tokenAmountToReceive: number;
  offer: Route;
};

export type ChainType = {
  chainId: number;
  chainName: string;
};

export type StepTransaction = {
  to?: string;
  data?: Hex;
  value?: bigint;
  gasLimit?: bigint;
  gasPrice?: bigint;
  chainId?: number;
  type?: number | string;
  transactionType?: StepType;
};

export type StepType = 'swap' | 'cross' | 'lifi' | 'custom' | 'approval';
