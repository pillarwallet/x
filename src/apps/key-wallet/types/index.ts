export interface Asset {
  id: number;
  name: string;
  symbol: string;
  logo: string;
  balance: number;
  decimals: number;
  price: number;
  price_change_24h: number;
  contract: string;
  chainId: number;
  chainName: string;
  usdBalance: number;
}

export interface SendFormData {
  amount: string;
  recipient: string;
}

export interface TransactionStatus {
  hash: string;
  chainId: number;
  status: 'pending' | 'success' | 'failed';
  timestamp: number;
}

