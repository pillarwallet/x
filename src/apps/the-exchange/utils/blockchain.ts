import { BigNumber, BigNumberish } from 'ethers';
import { formatEther, formatUnits } from 'ethers/lib/utils';
import { decodeFunctionData, erc20Abi } from 'viem';

// types
import { Token } from '../../../services/tokensData';
import { StepTransaction } from './types';

export const processBigNumber = (val: BigNumber): number =>
  Number(val.toString());

export const processEth = (val: BigNumberish, dec: number): number => {
  if (typeof val === 'bigint') {
    return +parseFloat(formatEther(val)).toFixed(2);
  }

  return +parseFloat(formatUnits(val as BigNumberish, dec));
};

// Utility: get native token symbol for a chain
export const NATIVE_SYMBOLS: Record<number, string> = {
  1: 'ETH',
  100: 'xDAI',
  137: 'POL',
  10: 'ETH',
  42161: 'ETH',
  56: 'BNB',
  8453: 'ETH',
};

// Helper: Detect if a tx is a native fee step
export const isNativeFeeTx = (
  tx: StepTransaction,
  feeReceiver: string
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof feeReceiver === 'string' &&
    tx.to.toLowerCase() === feeReceiver.toLowerCase()
  );
};

// Helper: Detect if a tx is an ERC20 (stablecoin or wrapped) fee step
export const isERC20FeeTx = (
  tx: StepTransaction,
  swapToken: Token
): boolean => {
  return (
    typeof tx.to === 'string' &&
    typeof swapToken.contract === 'string' &&
    tx.to.toLowerCase() === swapToken.contract.toLowerCase() &&
    tx.value === BigInt(0) &&
    typeof tx.data === 'string' &&
    tx.data.startsWith('0xa9059cbb')
  );
};

// Helper: Extract fee amount from tx
export const getFeeAmount = (
  tx: StepTransaction,
  swapToken: Token,
  decimals: number
): string => {
  if (tx.value && tx.data === '0x') {
    // Native
    return formatEther(tx.value);
  }
  if (isERC20FeeTx(tx, swapToken)) {
    try {
      const decoded = decodeFunctionData({
        abi: erc20Abi,
        data: tx.data || '0x',
      });
      if (decoded.args && typeof decoded.args[1] === 'bigint') {
        return formatUnits(decoded.args[1], decimals);
      }
    } catch (e) {
      return '0';
    }
  }
  return '0';
};

// Helper: Get fee symbol
export const getFeeSymbol = (
  tx: StepTransaction,
  swapToken: Token,
  chainId: number
): string => {
  if (tx.value && tx.data === '0x') {
    return NATIVE_SYMBOLS[chainId] || 'NATIVE';
  }
  return swapToken.symbol;
};
