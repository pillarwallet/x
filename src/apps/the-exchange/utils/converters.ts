export const hasThreeZerosAfterDecimal = (num: number): boolean => {
  const decimalPart = num.toString().split('.')[1] || '';
  return decimalPart.startsWith('000');
};

export const formatTokenAmount = (amount?: number) => {
  if (amount === undefined) return 0;
  return hasThreeZerosAfterDecimal(amount)
    ? amount.toFixed(8)
    : amount.toFixed(4);
};
