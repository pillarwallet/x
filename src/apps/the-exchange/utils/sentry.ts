import * as Sentry from '@sentry/react';
import { useWalletAddress } from '@etherspot/transaction-kit';

// Sentry configuration for the-exchange app
export const initSentryForExchange = () => {
  Sentry.setTag('app', 'the-exchange');
  Sentry.setTag('module', 'exchange');
};

// Utility to get fallback wallet address for logging
// This function should be called from within a React component context
// where the wallet address is available
export const fallbackWalletAddressForLogging = (): string => {
  // This is a utility function that should be called from within React components
  // The actual wallet address should be passed as a parameter or obtained via hook
  // For now, return unknown as this is a fallback utility
  return 'unknown_wallet_address';
};

// Enhanced Sentry logging with wallet address context
export const logExchangeEvent = (
  message: string,
  level: Sentry.SeverityLevel = 'info',
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel(level);
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_data', extra);
    }

    Sentry.captureMessage(message);
  });
};

// Log exchange errors with wallet address
export const logExchangeError = (
  error: Error | string,
  extra?: Record<string, unknown>,
  tags?: Record<string, string>
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.withScope((scope) => {
    scope.setLevel('error');
    scope.setTag('wallet_address', walletAddress);
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('error_type', 'exchange_error');

    if (tags) {
      Object.entries(tags).forEach(([key, value]) => {
        scope.setTag(key, value);
      });
    }

    if (extra) {
      scope.setExtra('exchange_error_data', extra);
    }

    if (error instanceof Error) {
      Sentry.captureException(error);
    } else {
      Sentry.captureMessage(error, 'error');
    }
  });
};

// Generic operation logging function
export const logOperation = (
  operationType: string,
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('operation_type', operationType);
    scope.setTag(`${operationType}_operation`, operation);

    scope.setExtra(`${operationType}_data`, data);

    Sentry.captureMessage(
      `${operationType.charAt(0).toUpperCase() + operationType.slice(1)} operation: ${operation}`,
      'info'
    );
  });
};

// Log swap operations
export const logSwapOperation = (
  operation: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('swap', operation, data, walletAddress);
};

// Log token operations
export const logTokenOperation = (
  operation: string,
  tokenData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('token', operation, tokenData, walletAddress);
};

// Log offer operations
export const logOfferOperation = (
  operation: string,
  offerData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('offer', operation, offerData, walletAddress);
};

// Log transaction operations
export const logTransactionOperation = (
  operation: string,
  transactionData: Record<string, unknown>,
  walletAddress?: string
) => {
  logOperation('transaction', operation, transactionData, walletAddress);
};

// Log user interactions
export const logUserInteraction = (
  interaction: string,
  data: Record<string, unknown>,
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('interaction_type', 'user_action');
    scope.setTag('user_interaction', interaction);

    scope.setExtra('interaction_data', data);

    Sentry.captureMessage(`User interaction: ${interaction}`, 'info');
  });
};

// Log performance metrics
export const logPerformanceMetric = (
  metric: string,
  value: number,
  unit: string = 'ms',
  walletAddress?: string
) => {
  Sentry.withScope((scope) => {
    scope.setLevel('info');
    scope.setTag(
      'wallet_address',
      walletAddress || fallbackWalletAddressForLogging()
    );
    scope.setTag('app_module', 'the-exchange');
    scope.setTag('metric_type', 'performance');
    scope.setTag('metric_name', metric);

    scope.setExtra('performance_data', {
      metric,
      value,
      unit,
      timestamp: new Date().toISOString(),
    });

    Sentry.captureMessage(
      `Performance metric: ${metric} = ${value}${unit}`,
      'info'
    );
  });
};

// Hook to get wallet address for logging
export const useWalletAddressForLogging = () => {
  const walletAddress = useWalletAddress();
  return walletAddress || 'unknown_wallet_address';
};

// Note: Error boundary removed due to TypeScript/JSX compatibility issues
// Use Sentry's default error boundary or create a separate React component file

// Transaction monitoring for exchange operations
export const startExchangeTransaction = (
  operation: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _data: Record<string, unknown> = {},
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _walletAddress?: string
) => {
  return Sentry.startSpan(
    {
      name: `exchange.${operation}`,
      op: 'exchange.operation',
    },
    (span) => {
      // Note: Span API has changed in Sentry v10
      // Properties are set via the span context instead
      return span;
    }
  );
};

// Breadcrumb utilities for exchange
export const addExchangeBreadcrumb = (
  message: string,
  category: string = 'exchange',
  data?: Record<string, unknown>,
  level: Sentry.SeverityLevel = 'info'
) => {
  const walletAddress = fallbackWalletAddressForLogging();

  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data: {
      ...data,
      wallet_address: walletAddress,
      app_module: 'the-exchange',
    },
  });
};
