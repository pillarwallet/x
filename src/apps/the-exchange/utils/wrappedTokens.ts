export const NATIVE_TOKEN_ADDRESSES = new Set([
  '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
  '0x0000000000000000000000000000000000000000',
]);

// Not including XDAI below
export const WRAPPED_NATIVE_TOKEN_ADDRESSES: Record<number, string> = {
  // Ethereum
  1: '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2', // WETH
  // Polygon
  137: '0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270', // WMATIC
  // Optimism
  10: '0x4200000000000000000000000000000000000006', // WETH
  // Arbitrum
  42161: '0x82aF49447D8a07e3bd95BD0d56f35241523fBab1', // WETH
  // Base
  8453: '0x4200000000000000000000000000000000000006', // WETH
  // BNB
  56: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c', // WBNB
};

export const isWrappedToken = (tokenAddress: string, chainId: number) => {
  return (
    tokenAddress.toLowerCase() ===
    WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId]?.toLowerCase()
  );
};

export const isNativeToken = (address: string) =>
  NATIVE_TOKEN_ADDRESSES.has(address.toLowerCase());

export const getWrappedTokenAddressIfNative = (
  tokenAddress: string,
  chainId: number
): string => {
  if (isNativeToken(tokenAddress)) {
    const wrappedAddress = WRAPPED_NATIVE_TOKEN_ADDRESSES[chainId];
    // fallback to token address in the case of XDAI token
    return wrappedAddress || tokenAddress;
  }
  return tokenAddress;
};
