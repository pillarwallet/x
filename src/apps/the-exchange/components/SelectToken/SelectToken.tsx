// types
import { CardPosition } from '../../utils/types';

// utils
import { getChainName } from '../../../../utils/blockchain';
import { limitDigits } from '../../../token-atlas/utils/converters';

// components
import Body from '../Typography/Body';

// images
import TokenLogo from '../TokenLogo/TokenLogo';

type SelectTokenProps = {
  type: CardPosition;
  tokenName?: string;
  tokenChain?: number;
  tokenLogo?: string;
  tokenBalance?: number;
  tokenPrice?: number;
  onClick?: () => void;
};

const SelectToken = ({
  type,
  tokenName,
  tokenChain,
  tokenLogo,
  tokenBalance,
  tokenPrice,
  onClick,
}: SelectTokenProps) => {
  const handleClick = () => {
    if (onClick) {
      onClick();
    }
  };

  return (
    <div
      id="select-token-info-exchange"
      onClick={handleClick}
      className="flex justify-between items-start mb-4"
    >
      <div className="flex w-full flex-col">
        <Body className="capitalize line-clamp-2 text-ellipsis break-all">
          {tokenName ?? type}
        </Body>
        <Body className="font-normal capitalize">
          {tokenChain ? `On ${getChainName(tokenChain)}` : 'Select Token'}
        </Body>
        {tokenBalance && (
          <Body className="font-normal">
            {limitDigits(tokenBalance)}{' '}
            {tokenPrice && (
              <span className="text-black_grey/[.4]">
                ${(tokenPrice * tokenBalance).toFixed(4)}
              </span>
            )}
          </Body>
        )}
      </div>
      <TokenLogo
        tokenName={tokenName}
        tokenLogo={tokenLogo}
        showLogo={Boolean(tokenName)}
      />
    </div>
  );
};

export default SelectToken;
