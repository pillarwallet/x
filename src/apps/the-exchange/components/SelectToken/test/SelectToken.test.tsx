import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// test utils
import { ExchangeTestWrapper } from '../../../../../test-utils/testUtils';

// types
import { CardPosition } from '../../../utils/types';

// components
import SelectToken from '../SelectToken';

describe('<SelectToken />', () => {
  const mockOnClick = vi.fn();
  const type = CardPosition.SWAP;
  const tokenName = 'Ether';
  const tokenChain = 1;
  const tokenLogo = 'iconEth.png';

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <ExchangeTestWrapper>
          <SelectToken
            type={type}
            tokenName={tokenName}
            tokenChain={tokenChain}
            tokenLogo={tokenLogo}
            onClick={mockOnClick}
          />
        </ExchangeTestWrapper>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders SelectToken correctly with arguments', () => {
    render(
      <ExchangeTestWrapper>
        <SelectToken
          type={type}
          tokenName={tokenName}
          tokenChain={tokenChain}
          tokenLogo={tokenLogo}
          onClick={mockOnClick}
        />
      </ExchangeTestWrapper>
    );

    expect(screen.getByText('Ether')).toBeInTheDocument();
    expect(screen.getByText('On Ethereum')).toBeInTheDocument();
  });

  it('renders "Select Token" with type when token is not provided', () => {
    render(
      <ExchangeTestWrapper>
        <SelectToken type={type} onClick={mockOnClick} />
      </ExchangeTestWrapper>
    );

    expect(screen.getByText('Select Token')).toBeInTheDocument();
  });

  it('calls onClick when SelectToken is clicked', () => {
    render(
      <ExchangeTestWrapper>
        <SelectToken
          type={type}
          tokenName={tokenName}
          tokenChain={tokenChain}
          tokenLogo={tokenLogo}
          onClick={mockOnClick}
        />
      </ExchangeTestWrapper>
    );

    const selectTokenElement = screen.getByText('Ether');
    fireEvent.click(selectTokenElement);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });
});
