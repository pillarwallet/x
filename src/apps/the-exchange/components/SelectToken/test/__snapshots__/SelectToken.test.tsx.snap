// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SelectToken /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex justify-between items-start mb-4"
  id="select-token-info-exchange"
  onClick={[Function]}
>
  <div
    className="flex w-full flex-col"
  >
    <p
      className="text-base font-medium text-black_grey capitalize line-clamp-2 text-ellipsis break-all"
    >
      Ether
    </p>
    <p
      className="text-base font-medium text-black_grey font-normal capitalize"
    >
      On Ethereum
    </p>
  </div>
  <div
    className="flex min-w-fit items-end"
  >
    <div
      className="w-5 h-5 relative rounded-full grayscale"
    >
      <img
        alt="token-logo"
        className="w-full h-full object-fill rounded-full"
        onError={[Function]}
        src="iconEth.png"
      />
    </div>
    <div
      className="w-3 h-3 rounded-full border border-white relative right-[7px] hidden"
    >
      <img
        alt="chain-logo"
        className="w-full h-full object-fill rounded-full grayscale hidden"
      />
    </div>
  </div>
</div>
`;
